### ISSUE TITLE
Can't seem to update RITM Custom Variable field

### ISSUE THREAD

**Issue by jbaechtelMT** [insomniacc/PSSnow#43]
I have a Custom Variable named "**subject**".  I am able to read the variable in the **REST API** (both Postman & API Explorer) using the name "**variables.subject**".  When I try to update it through the REST API, it never updates.

**I can successfully update native RITM variables** ("**short_description**" & "**work_notes**") but not any Custom Variables.

Does anybody have a successful example of being able to update Custom Variables using REST API?

I also tried updating them using the **Set-SNOWSCRequestedItem** function in **PSSnow** using the "**-Properties**" parameter but no luck.  Any ideas?

Thanks,

**Comment by insomniacc** [comment]
Okay, I've had chance to look into this. I think I understand what's happening here.
Until your post I wasn't actually aware that you could access variables of a RITM simply via the -fields param using dot walking.
It seems this is a quality of life/ease of use thing that servicenow have implemented.

Infact I actually created a function to get variables with Get-SNOWRITMVariableSet (which is still required if you don't know the variable names).

Ultimately though, those properties that you're accessing, you are dot walking to another table record, they are not contained on the 'sc_req_item' table where your RITM record is.
Dot walking is only available when using GET commands and is not supported by SET.

Variables are quite confusing in the back end, there's actually 3 tables involved:
sc_item_option_mtom (Pointer records that link the below table records to the RITM itself)
sc_item_option (value of the variable)
item_option_new (display value / question of the variable field)

Basically, if you want to SET/Update your variable, you'd have to do this within sc_item_option table.

Let me give you an example, I have RITM0010001, with the following variables:

- primary_contact  : 1b3f2d6047292110d3e5fa8bd36d4351
- cost_center      : a581ab703710200044e0bfc8bcbe5de8
- ip_range         : 127.0.0.1
- business_purpose : testing

I'm going to update business_purpose to say "hello world".

```powershell
#First I get all the mtom pointers for my RITM
$MTOMRecords = Get-SNOWObject -Table "sc_item_option_mtom" -Query "request_itemLIKERITM0010001" -Fields @( "sc_item_option.item_option_new.name", "sc_item_option", "sc_item_option.value") -ExcludeReferenceLinks
#I find the one thats relevant to the variable I want to update
$MTOMRecord = $MTOMRecords | ?{$_."sc_item_option.item_option_new.name" -eq "business_purpose"}
#I update the variable's value in the sc_item_option table
Set-SNOWObject -table "sc_item_option" -sys_id $MTOMRecord.sc_item_option -Properties @{value="Hello World"}
```

It's a bit of a faff so I might look at adding a function to make this easier in future, but figured there probably wasn't much call for changing these variables after the fact. I'm curious though, what's your use case here?

**Comment by insomniacc** [comment]
Add: Set-SNOWRITMVariable 
Summary: Provide a sys_id or number, variable sys_id or name, and a new value, as an easy option to update the sc_item_option records.

I'll get this added in 1.1.0

**Comment by jbaechtelMT** [comment]
Thanks for looking into this.

Use Case:

   - Tickets are created in ServiceNow with Custom Variables populated
   - Automation scripts (outside of SN) read the tickets and get info from
   Custom Variable fields
   - Scripts perform the work being requested (ex:  Create new AD Groups
   and add AD Users into New Group)
   - Update SN ticket (updating Custom Variables fields) with results of
   script's output
   - Optionally:  Close SN ticket

Thanks again,

Joe

**Comment by insomniacc** [comment]
No prob! I've added Set-SNOWRITMVariable in the last commit, this will be available in the next release.

**Comment by insomniacc** [comment]
@jbaechtelMT - It's been published on the PSGallery, v1.1.0 if you want to use it! 👍🏻
